### ISSUE TITLE
Updated Entities not being Stitched Due to Hashing

### ISSUE THREAD

**Issue by ryanaross** [backstage/backstage#7199]
In this change #6827 (@freben), hashing was introduced to reduce the number of times entities are stitched. 

We are seeing an issue where an entity is updated, with the removal of the `catalog-info.yaml`, but the hash is not updated and therefore it is not stitched again. Without stitching, the entity is never marked as orphaned.

## Expected Behavior

The hash should be getting updated with the error we are seeing in the logs for this entity, after the catalog-info.yaml has been removed:
```
Unable to read url, NotFoundError: https://github.com/org/repo-name/blob/main/catalog-info.yaml could not be read as https://api.github.com/repos/org/repo-name/contents/catalog-info.yaml?ref=main, 404 Not Found type=plugin entity=location:default/generated-a80...
```

And then it should be re-stitched, and its orphan status would be found. But the above error is only being discovered later in the processing. 

If I comment out the [hash comparison block](https://github.com/backstage/backstage/blob/7ab55167d047efb60114cca32af39feacbea89ee/plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts#L174), items are re-stitched and the orphan status is quickly discovered.

## Current Behavior

Updated entities, specifically updated with the removal of catalog-info.yaml, are not being marked as orphaned.

## Possible Solutions

- Wait for a possible 404 error to be discovered before hashing.
- Make hashing optional

## Steps to Reproduce

1. Configure github location discovery to find a catalog-info.yaml for a component upstream in a github.com repository
2. Remove catalog-info.yaml from the repository.
3. Observe that orphan status is never set to true. (`metadata.annotations: "backstage.io/orphan": "true"`)

## Context

We are starting to accumulate deleted items that need to be removed from the Backstage catalog, but discovering them is difficult.

## Your Environment

release-2021-08-26 , deployed via helm/container. Can reproduce locally.

**Comment by Rugvip** [comment]
Thank you for reporting, good catch! 👍 

I think this might be as simple as `errors` (and `ok`?) being part of the hashing.
Maybe it's something you'd like to look at contributing? Otherwise we'll be able to prioritize it too

**Comment by ryanaross** [comment]
> Thank you for reporting, good catch! 👍
> 
> I think this might be as simple as `errors` (and `ok`?) being part of the hashing.
> Maybe it's something you'd like to look at contributing? Otherwise we'll be able to prioritize it too

Ah I would like to, but lack some required knowledge. I just tried to stringify the results.ok and put it in the hash but it did not work, and I am not sure why, that seemed like it would.

The hash is including [errorStrings in the hash](https://github.com/backstage/backstage/blob/7ab55167d047efb60114cca32af39feacbea89ee/plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts#L164) currently. But this particular error only appears 10 seconds **after** the hash has been created, so it cannot be included I think.

**Comment by Rugvip** [comment]
Ah yeah sorry that's just me being confused. This is the actually intended behavior, where we don't want to delete or mark the entity as orphaned if reading the parent location fails with an error. It's however not intended for those entities to be difficult to discover, so I think we can consider this to be a duplicate of #6973

**Comment by ryanaross** [comment]
Thanks for that Issue link @Rugvip, I hadn't seen it. Oddly, if I remove the hashing mechanism, these entities whose repositories have been deleted do get marked as orphaned the next time they are processed.

**Comment by Rugvip** [comment]
Ah I think I know what's going on here. There's likely a bug or mismatch of expectations in the catalog processing that was causing the processing of locations to be considered successful even if it had errors.

We assume that if the processing does not throw any errors then it's ok, but we might have errors being emitted along the way, which is what our reader for GitHub URLs is doing when it fails.

https://github.com/backstage/backstage/blob/81c958ad14bb3a5a678e63db3dbf73dc4e10c505/plugins/catalog-backend/src/next/processing/DefaultCatalogProcessingOrchestrator.ts#L124

https://github.com/backstage/backstage/blob/7ab55167d047efb60114cca32af39feacbea89ee/plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.ts#L78

@freben @jhaals @benjdlambert I don't think we've talked about handling different kinds of errors differently right? As in thrown vs emitted. I suspect the fix here is to `ok: emitted.errors.length === 0` or something of that sort

**Comment by freben** [comment]
Hm interesting. Well the semantics of `ok` is currently to differentiate between whether it threw or not and we can probably keep that, I think it's the caller that needs to have that errors checker right? And it kind of does.

Is it the case that we should throw NotFound instead of emitting it? Have to look closer at it tomorrow

**Comment by Rugvip** [comment]
Yeah could keep the way `ok` works, but is it perhaps a bit surprising? Should make sure it's documented well in the processing API in that case as it's a pretty big difference. ​I agree we should probably fix this by changing the URL reader but it might be that we'll just run into more issues like this in the future then.

Perhaps a fix is to remove errors as a thing you can emit and replace them with warnings instead? And just saying that any emitted error is actually a warning for compatibility. That way I think it's a bit more clear what the behaviour is from and API point of view? Even though it's really just the same as what we have atm.

**Comment by freben** [comment]
It'd probably break third party processors. I'd envisioned emitting actual warnings as a different thing than errors, in the future. But I get the feeling that we need some slightly richer signalling here overall... The difference between "I exploded violently, abort everything", vs "I am working correctly but the data was bad, keep processing my children and assume they didn't change", vs "One of the two children I usually emit has now vanished but the other one is fine", vs ... yeah.

Anyway yeah this needs some drawing board sketching to make sure that we get it right.

**Comment by Rugvip** [comment]
This was fixed by #7413

**Comment by Rugvip** [comment]
@kim5566 that is intended. The intent of the design is that entities should only be marked as orphaned if we know that they are truly missing, not if there are processing errors.
